nixos/modules/hardware/device-tree.nix

```diff
@@ -4,6 +4,100 @@ with lib;
 
 let
   cfg = config.hardware.deviceTree;
+
+  overlayType = types.submodule {
+    options = {
+      name = mkOption {
+        type = types.str;
+        description = ''
+          Name of this overlay
+        '';
+      };
+
+      dtsFile = mkOption {
+        type = types.nullOr types.path;
+        description = ''
+          Path to .dts overlay file, overlay is applied to
+          each .dtb file matching "compatible" of the overlay.
+        '';
+        default = null;
+        example = literalExample "./dts/overlays.dts";
+      };
+
+      dtsText = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description = ''
+          Literal DTS contents, overlay is applied to
+          each .dtb file matching "compatible" of the overlay.
+        '';
+        example = literalExample ''
+          /dts-v1/;
+          /plugin/;
+          / {
+                  compatible = "raspberrypi";
+                  fragment@0 {
+                          target-path = "/soc";
+                          __overlay__ {
+                                  pps {
+                                          compatible = "pps-gpio";
+                                          status = "okay";
+                                  };
+                          };
+                  };
+          };
+        '';
+      };
+
+      dtboFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = ''
+          Path to .dtbo compiled overlay file.
+        '';
+      };
+    };
+  };
+
+  # this requires kernel package
+  dtbsWithSymbols = pkgs.stdenv.mkDerivation {
+        name = "dtbs-with-symbols";
+        inherit (cfg.kernelPackage) src nativeBuildInputs;
+        patches = map (patch: patch.patch) cfg.kernelPackage.kernelPatches;
+
+        buildPhase = ''
+          substituteInPlace scripts/Makefile.lib \
+            --replace 'DTC_FLAGS += $(DTC_FLAGS_$(basetarget))' 'DTC_FLAGS += $(DTC_FLAGS_$(basetarget)) -@'
+          make ${pkgs.stdenv.hostPlatform.platform.kernelBaseConfig}
+          make dtbs
+        '';
+
+        installPhase = ''
+          make dtbs_install INSTALL_DTBS_PATH=$out/dtbs
+        '';
+      };
```

Comment:
Would it make sense to move this function, and maybe also `filterDTBs` and `compileDTS` to nixpkgs (next to `applyOverlays`)? I think these functions might sometimes be useful for users to build their own device tree packages, and having them inside the module makes it hard to reuse.

Reply:
`compileDTS` sounds good, not sure about `filterDTBs` as that is a pretty generic store path filter which resembles `filterSource` but for store paths..

Reply:
For now I'm keeping this internal, if you would like to have this exposed from `pkgs` let me know and I'll try to fit it somewhere but I'm bad at namespaces and naming :) It also needs a better name in that case, something like `pkgs.compileDeviceTreeOverlays`.